### PR TITLE
Fixed #30062 -- Added support for unique conditional constraints.

### PR DESCRIPTION
**Pull request by tbicr**
related to https://github.com/django/django/pull/10789 refactoring, so PR changes only in 
`add unique conditional constraint` commit

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -1063,6 +1063,8 @@ def _perform_unique_checks(self, unique_checks):
                 continue
 
             qs = model_class._default_manager.filter(**lookup_kwargs)
+            if condition:
+                qs = qs.filter(condition)
```
Comment (charettes) [inline django/db/models/base.py]: That's too naive. It has to consider whether or not `self` matches the condition in the first place.

File: django/db/models/constraints.py [hunk 2]
```diff
@@ -66,34 +67,56 @@ def deconstruct(self):
 
 
 class UniqueConstraint(BaseConstraint):
-    def __init__(self, *, fields, name):
+    def __init__(self, *, fields, name, condition=None):
+        if not isinstance(fields, (list, tuple)):
+            raise ValueError('UniqueConstraint.fields must be a list or tuple.')
```
[the remark this answers is not in the document]
Comment (tbicr) [reply django/db/models/constraints.py]: removed tests and checks

File: django/db/models/constraints.py [hunk 3]
```diff
@@ -66,34 +67,56 @@ def deconstruct(self):
...
 
     def create_sql(self, model, schema_editor):
         fields = [model._meta.get_field(field_name).column for field_name in self.fields]
-        return schema_editor._create_unique_sql(model, fields, self.name)
+        condition = self._get_condition_sql(model, schema_editor)
+        return schema_editor._create_unique_sql(model, fields, self.name, condition=condition)
 
     def remove_sql(self, model, schema_editor):
-        return schema_editor._delete_unique_sql(model, self.name)
+        return schema_editor._delete_unique_sql(model, self.name, bool(self.condition))
```
[the remark this answers is not in the document]
Comment (tbicr) [reply django/db/models/constraints.py]: changed

File: django/forms/models.py [hunk 4]
```diff
@@ -687,7 +687,9 @@ def validate_unique(self):
 
         errors = []
         # Do each of the unique checks (unique and unique_together)
-        for uclass, unique_check in all_unique_checks:
+        for uclass, unique_check, condition in all_unique_checks:
+            if condition:
+                continue
```
[the remark this answers is not in the document]
Comment (tbicr) [reply django/forms/models.py]: removed, see comment below

File: django/db/models/base.py (hunk 1 above)
Comment (tbicr) [reply django/db/models/base.py]: removed this check at all, left only db constraint check

File: django/db/models/base.py [hunk 5]
```diff
@@ -1004,13 +1004,13 @@ def _get_unique_checks(self, exclude=None):
...
                     # Add the check if the field isn't excluded.
-                    unique_checks.append((model_class, tuple(check)))
+                    unique_checks.append((model_class, tuple(check), None))
 
         for model_class, model_constraints in constraints:
             for constraint in model_constraints:
                 if (isinstance(constraint, UniqueConstraint) and
                         not any(name in exclude for name in constraint.fields)):
-                    unique_checks.append((model_class, constraint.fields))
+                    unique_checks.append((model_class, constraint.fields, constraint.condition))
```
[the remark this answers is not in the document]
Comment (tbicr) [reply django/db/models/base.py]: Cool explanation, removed conditional uniqueness check in code: https://github.com/django/django/pull/10796/commits/ca11bd17d1da224573f3470ce23515be608f6c5b#diff-507b415116b409afa4f723e41a759a9eR1012

File: tests/constraints/tests.py [hunk 6]
```diff
@@ -68,21 +84,99 @@ def test_repr(self):
...
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIn(constraint.name, get_constraints(Product._meta.db_table))
+            editor.remove_constraint(Product, constraint)
+        self.assertNotIn(constraint.name, get_constraints(Product._meta.db_table))
+
+    def test_database_constraint_with_condition(self):
+        Product.objects.create(name='p1')
+        Product.objects.create(name='p1', color='red')
```
Comment (charettes) [inline tests/constraints/tests.py]: Could use `bulk_create`

```suggestion
        Product.objects.bulk_create([
            Product(name='p1'),
            Product(name='p1', color='red'),
        ])
```

File: tests/constraints/tests.py [hunk 7]
```diff
@@ -68,21 +84,99 @@ def test_repr(self):
...
+    def test_delete_conditional_constraint(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIn(constraint.name, get_constraints(Product._meta.db_table))
+            editor.remove_constraint(Product, constraint)
+        self.assertNotIn(constraint.name, get_constraints(Product._meta.db_table))
```
Comment (charettes) [inline tests/constraints/tests.py]: Unless the constraint is added back this alters the global test state. Most of the schema altering tests above are also affected by that. These tests should live in the `schema` and `migrations` test apps which has the necessary setup in place to prevent this from happening.

Have a look at #7615 and #10639 for examples.

File: tests/constraints/tests.py (hunk 6 above)
Comment (tbicr) [reply tests/constraints/tests.py]: changed

File: tests/constraints/tests.py (hunk 7 above)
Comment (tbicr) [reply tests/constraints/tests.py]: refactored and moved to tests/migrations/test_operations.py

**Comment by tbicr** [comment]
Just to check: are any additional changes need for this MR? It's cool to have unique conditional constraint in 2.2 (in addition to unique constraint and conditional indexes), I have time now to finish this MR.

**Comment by pope1ni** [comment]
@tbicr It looks as though you have forgotten to uncheck _Patch needs improvement_ on the [ticket](https://code.djangoproject.com/ticket/30062), so it isn't on the review queue. You should probably also assign the ticket to yourself.

File: docs/ref/models/constraints.txt [hunk 8]
```diff
@@ -68,3 +68,41 @@ ensures only one location can exist for each ``date``.
...
+
+For example ``UniqueConstraint(fields=['email'], condition=Q(status='ACTIVE')``
+ensures your emails in ACTIVE status are unique.
+
+.. admonition:: Restrictions on PostgreSQL
+
+    PostgreSQL requires functions referenced in the condition to be be marked as
+    IMMUTABLE. Django doesn't validate this but PostgreSQL will error. This
+    means that functions such as :ref:`date-functions` and
+    :class:`~django.db.models.functions.Concat` aren't accepted. If you store
```
Comment (timgraham) [inline docs/ref/models/constraints.txt]: I guess this note was copied from ref/models/indexes.txt but I'm not clear on why database functions can be used with Q objects (perhaps as lookups?).

File: tests/constraints/tests.py [hunk 9]
```diff
@@ -80,21 +83,38 @@ def test_repr(self):
     def test_deconstruction(self):
         fields = ['foo', 'bar']
         name = 'unique_fields'
-        check = models.UniqueConstraint(fields=fields, name=name)
-        path, args, kwargs = check.deconstruct()
+        constraint = models.UniqueConstraint(fields=fields, name=name)
```
Comment (timgraham) [inline tests/constraints/tests.py]: Would be better to do unrelated cleanups like this (I saw another rename in schema.py) in another commit.

File: tests/constraints/tests.py [hunk 10]
```diff
@@ -80,21 +83,38 @@ def test_repr(self):
...
+        path, args, kwargs = constraint.deconstruct()
         self.assertEqual(path, 'django.db.models.UniqueConstraint')
         self.assertEqual(args, ())
         self.assertEqual(kwargs, {'fields': tuple(fields), 'name': name})
 
+    def test_deconstruction_with_condition(self):
+        fields = ['foo', 'bar']
+        name = 'unique_fields'
+        constraint = models.UniqueConstraint(fields=fields, name=name, condition=models.Q(foo=models.F('bar')))
```
Comment (timgraham) [inline tests/constraints/tests.py]: Did you omit a variable for `condition` (unlike `fields` and `name`) for some reason?

File: tests/constraints/tests.py [hunk 11]
```diff
@@ -80,21 +83,38 @@ def test_repr(self):
...
 
     def test_model_validation(self):
-        with self.assertRaisesMessage(ValidationError, 'Product with this Name already exists.'):
-            Product(name=self.p1.name).validate_unique()
+        with self.assertRaisesMessage(ValidationError, 'Product with this Name and Color already exists.'):
+            Product(name=self.p1.name, color=self.p1.color).validate_unique()
+
+    def test_model_validation_with_condition(self):
+        Product(name=self.p1.name, color='blue').validate_unique()
+        Product(name=self.p2.name).validate_unique()  # validation ignored, will raise on DB level
```
Comment (timgraham) [inline tests/constraints/tests.py]: I think comment would be better as a docstring:
`"""Partial unique constraints are ignored by Model.validate_unique()."""`

File: django/db/backends/base/schema.py [hunk 12]
```diff
@@ -1027,17 +1039,32 @@ def create_unique_name(*args, **kwargs):
...
-    def _delete_unique_sql(self, model, name):
+    def _delete_unique_sql(self, model, name, condition=None):
+        if condition:
+            if not self.connection.features.supports_partial_indexes:
+                return None
+            return self._delete_constraint_sql(self.sql_delete_index, model, name)
         return self._delete_constraint_sql(self.sql_delete_unique, model, name)
 
-    def _check_sql(self, name, check):
+    def _check_sql(self, model, name, check):
```
Comment (timgraham) [inline django/db/backends/base/schema.py]: Accidentally added this unused argument?

File: django/db/models/constraints.py [hunk 13]
```diff
@@ -66,34 +67,55 @@ def deconstruct(self):
...
     def remove_sql(self, model, schema_editor):
-        return schema_editor._delete_unique_sql(model, self.name)
+        condition = self._get_condition_sql(model, schema_editor)
+        return schema_editor._delete_unique_sql(model, self.name, condition=condition)
 
     def __repr__(self):
-        return '<%s: fields=%r name=%r>' % (self.__class__.__name__, self.fields, self.name)
+        return '<%s: fields=%r name=%r%s>' % (
+            self.__class__.__name__, self.fields, self.name,
+            '' if self.condition is None else ', condition=%s' % self.condition,
```
Comment (timgraham) [inline django/db/models/constraints.py]: Please check test coverage carefully. I didn't spot a test for this change.

File: docs/ref/models/constraints.txt (hunk 8 above)
Comment (tbicr) [reply docs/ref/models/constraints.txt]: I can provide next abstract example: `models.UniqueConstraint(fields=['z1'], name='z_uniq', condition=models.Q(z1=Lower('z2')))`, eg. when you use it with two fields.

File: tests/constraints/tests.py (hunk 9 above)
Comment (tbicr) [reply tests/constraints/tests.py]: removed, should I add new commit in this MR or other?

Comment (timgraham) [reply tests/constraints/tests.py]: sure, you can add it here.

File: tests/constraints/tests.py (hunk 10 above)
Comment (tbicr) [reply tests/constraints/tests.py]: no reason, add `condition` variable like `fields` and `name`

File: tests/constraints/tests.py (hunk 11 above)
Comment (tbicr) [reply tests/constraints/tests.py]: changed

File: django/db/backends/base/schema.py (hunk 12 above)
Comment (tbicr) [reply django/db/backends/base/schema.py]: yep, removed

File: docs/ref/models/constraints.txt (hunk 8 above)
Comment (timgraham) [reply docs/ref/models/constraints.txt]: Okay. Maybe it would be better to something like "The same database restrictions that apply to Index.condition apply to UniqueConstraint.condition." (rather than repeating the same 4 boxes)

File: django/db/models/constraints.py (hunk 13 above)
Comment (tbicr) [reply django/db/models/constraints.py]: good point, added test, a bit changed `__repr__` for conditional constraint

File: docs/ref/models/constraints.txt (hunk 8 above)
Comment (tbicr) [reply docs/ref/models/constraints.txt]: changed to `UniqueConstraint.condition has the same database restrictions that Index.condition has.`

File: tests/constraints/tests.py (hunk 9 above)
Comment (tbicr) [reply tests/constraints/tests.py]: added as new commit

File: docs/ref/models/constraints.txt [hunk 14]
```diff
@@ -69,3 +69,17 @@ date.
...
+
+``condition``
+-------------
+
+.. attribute:: UniqueConstraint.condition
+
+A :class:`Q` object that specifies the condition you want the constraint to
+enforce.
+
+For example, ``UniqueConstraint(fields=['email'], condition=Q(status='ACTIVE')``
```
Comment (timgraham) [inline docs/ref/models/constraints.txt]: I'm having trouble wrapping my head around this example. What model has these fields and how does this constraint make logical sense? Perhaps we can find a more obvious example.

Comment (tbicr) [reply docs/ref/models/constraints.txt]: I think about next:

    class ResetEmail(models.Model):
        email = models.EmailField()
        status = model.CharField(choices=('ACTIVE', 'DONE', 'OUTDATED'))
        token =  model.CharField()
        class Meta:
            constraints = (
                # only one token can be used for email resetting at one time
                UniqueConstraint(fields=['email'], condition=Q(status='ACTIVE')),
            )

I also have another one:

    class DocumentBlock(models.Model):
        document = model.ForeignKey('Document')
        block_type = model.CharField(choices=('COVER', 'PARAGRAPH'))
        class Meta:
            constraints = (
                # document can have only one cover block
                UniqueConstraint(fields=['document'], condition=Q(block_type='COVER')),
            )

it's can be quite specific cases, so maybe someone can propose more clear example?

File: django/db/backends/base/schema.py [hunk 15]
```diff
@@ -1008,7 +1013,14 @@ def create_fk_name(*args, **kwargs):
     def _delete_fk_sql(self, model, name):
         return self._delete_constraint_sql(self.sql_delete_fk, model, name)
 
-    def _unique_sql(self, fields, name):
+    def _unique_sql(self, model, fields, name, condition=''):
```
Comment (timgraham) [inline django/db/backends/base/schema.py]: Should it be `condition=None` or `condition=''` in all signatures (currently some inconsistency)? I'd lean toward the former.

File: django/db/models/constraints.py [hunk 16]
```diff
@@ -66,34 +67,55 @@ def deconstruct(self):
...
+            '' if self.condition is None else ' condition=%s' % self.condition,
+        )
 
     def __eq__(self, other):
         return (
             isinstance(other, UniqueConstraint) and
             self.name == other.name and
-            self.fields == other.fields
+            self.fields == other.fields and
+            self.condition == other.condition
```
Comment (timgraham) [inline django/db/models/constraints.py]: Missing test for this line?

File: docs/ref/models/constraints.txt (hunk 14 above)
Comment (timgraham) [reply docs/ref/models/constraints.txt]: How about:
```
For example, ``UniqueConstraint(fields=['user'], condition=Q(status='DRAFT')``
ensures that each user only has one draft.
```

File: tests/migrations/test_operations.py [hunk 17]
```diff
@@ -501,6 +501,56 @@ def test_create_model_with_constraint(self):
...
+
+        # Test database alteration
+        self.assertTableNotExists("test_crmo_pony")
+        with connection.schema_editor() as editor:
+            operation.database_forwards("test_crmo", editor, project_state, new_state)
+        self.assertTableExists("test_crmo_pony")
+
+        # Test constraint actually works
+        with connection.cursor() as cursor:
+            cursor.execute("INSERT INTO test_crmo_pony (id, pink, weight) VALUES (1, 1, 4.0)")
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: Did you try the ORM rather than raw SQL? I'm not sure if there's a reason the test above doesn't use it.

File: django/db/backends/base/schema.py (hunk 15 above)
Comment (tbicr) [reply django/db/backends/base/schema.py]: Same `condition=''` for indexes: https://github.com/django/django/blob/master/django/db/backends/base/schema.py#L897

Comment (tbicr) [reply django/db/backends/base/schema.py]: I like `condition=None`, but I like more consistency with index signature, should I fix both unique_constraint and index signatures?

File: docs/ref/models/constraints.txt (hunk 14 above)
Comment (tbicr) [reply docs/ref/models/constraints.txt]: fine for me

File: django/db/backends/base/schema.py (hunk 15 above)
Comment (timgraham) [reply django/db/backends/base/schema.py]: Okay, as before, the other cleanups can be in a separate commit.

File: tests/migrations/test_operations.py (hunk 17 above)
Comment (tbicr) [reply tests/migrations/test_operations.py]: changed to ORM

File: django/db/models/constraints.py (hunk 16 above)
Comment (tbicr) [reply django/db/models/constraints.py]: added tests for `__eq__`

File: django/db/backends/base/schema.py (hunk 15 above)
Comment (tbicr) [reply django/db/backends/base/schema.py]: added as separate commit both for index and unique constraint

**Comment by timgraham** [comment]
It looks like a test isn't cleaning up after itself. Can you investigate?
```
$ ./tests/runtests.py migrations --parallel=1 --reverse --failfast
...
======================================================================
ERROR: test_create_model_with_partial_unique_constraint (migrations.test_operations.OperationTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/sqlite3/base.py", line 376, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.OperationalError: index test_constraint_pony_pink_for_weight_gt_5_uniq already exists

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tim/code/django/tests/migrations/test_operations.py", line 527, in test_create_model_with_partial_unique_constraint
    operation.database_forwards('test_crmo', editor, project_state, new_state)
  File "/home/tim/code/django/django/db/backends/sqlite3/schema.py", line 35, in __exit__
    super().__exit__(exc_type, exc_value, traceback)
  File "/home/tim/code/django/django/db/backends/base/schema.py", line 110, in __exit__
    self.execute(sql)
  File "/home/tim/code/django/django/db/backends/base/schema.py", line 137, in execute
    cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/tim/code/django/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/tim/code/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/tim/code/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/sqlite3/base.py", line 376, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.OperationalError: index test_constraint_pony_pink_for_weight_gt_5_uniq already exists
```

**Comment by tbicr** [comment]
@timgraham , sure, I will in next two days
